core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1047,17 +725,543 @@ private static SortOrder freshSortOrder(int orderId, Schema schema, SortOrder so
+      }
+
+      int maxSize = Math.max(1, PropertyUtil.propertyAsInt(properties,
+          TableProperties.METADATA_PREVIOUS_VERSIONS_MAX, TableProperties.METADATA_PREVIOUS_VERSIONS_MAX_DEFAULT));
+
+      List<MetadataLogEntry> newMetadataLog;
+      if (previousFiles.size() >= maxSize) {
+        int removeIndex = previousFiles.size() - maxSize + 1;
+        newMetadataLog = Lists.newArrayList(previousFiles.subList(removeIndex, previousFiles.size()));
+      } else {
+        newMetadataLog = Lists.newArrayList(previousFiles);
+      }
+      newMetadataLog.add(new MetadataLogEntry(timestampMillis, previousFileLocation));
+
+      return newMetadataLog;
+    }
+
+    /**
+     * Finds intermediate snapshots that have not been committed as the current snapshot.
+     *
+     * @return a set of snapshot ids for all added snapshots that were later replaced as the current snapshot in changes
+     */
+    private static Set<Long> intermediateSnapshotIdSet(List<MetadataUpdate> changes, long currentSnapshotId) {
+      Set<Long> addedSnapshotIds = Sets.newHashSet();
+      Set<Long> intermediateSnapshotIds = Sets.newHashSet();
+      for (MetadataUpdate update : changes) {
+        if (update instanceof MetadataUpdate.AddSnapshot) {
+          // adds must always come before set current snapshot
+          MetadataUpdate.AddSnapshot addSnapshot = (MetadataUpdate.AddSnapshot) update;
+          addedSnapshotIds.add(addSnapshot.snapshot().snapshotId());
+        } else if (update instanceof MetadataUpdate.SetCurrentSnapshot) {
+          long snapshotId = ((MetadataUpdate.SetCurrentSnapshot) update).snapshotId();
+          if (addedSnapshotIds.contains(snapshotId) && snapshotId != currentSnapshotId) {
+            intermediateSnapshotIds.add(snapshotId);
+          }
+        }
+      }
+
+      return intermediateSnapshotIds;
+    }
+
+    private static List<HistoryEntry> updateSnapshotLog(
+        List<HistoryEntry> snapshotLog, Map<Long, Snapshot> snapshotsById, long currentSnapshotId,
+        List<MetadataUpdate> changes) {
+      // find intermediate snapshots to suppress incorrect entries in the snapshot log.
+      //
+      // transactions can create snapshots that are never the current snapshot because several changes are combined
+      // by the transaction into one table metadata update. when each intermediate snapshot is added to table metadata,
+      // it is added to the snapshot log, assuming that it will be the current snapshot. when there are multiple
+      // snapshot updates, the log must be corrected by suppressing the intermediate snapshot entries.
+      //
+      // a snapshot is an intermediate snapshot if it was added but is not the current snapshot.
+      Set<Long> intermediateSnapshotIds = intermediateSnapshotIdSet(changes, currentSnapshotId);
+
+      // update the snapshot log
+      List<HistoryEntry> newSnapshotLog = Lists.newArrayList();
+      for (HistoryEntry logEntry : snapshotLog) {
```

Comment:
The only update to this block is checking whether a snapshot was intermediate and suppressing the log for it.